### PR TITLE
Ftr 534 webhook reverse notifications

### LINKED ISSUE #534: Reversed webhook notifications
When Pyramus fires webhook notifications of modified entities, the notifications are sent in a reverse order. For example, if you create a new student group and add staff members and students to it, notifications are sent in order...

STUDENTGROUP_STUDENT_CREATE
STUDENTGROUP_STAFFMEMBER_CREATE
STUDENTGROUP_CREATE

...rather than...

STUDENTGROUP_CREATE
STUDENTGROUP_STAFFMEMBER_CREATE
STUDENTGROUP_STUDENT_CREATE

...which causes synchronization problems in Muikku due to the fact that it is informed of a student group student having been created before being informed of the student group itself having been created.

### PR DESCRIPTION
**Pull request by anttiviljakainen**
* Migrated WebhookController to Webhooks
* Moved event handlers to their own class
* Added TransactionScoped queue
* Changed student group events to queue up to ensure proper order

Closes #534

### CODE REVIEW

File: pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  public synchronized void onCourseStaffMemberUpdatedAfterSuccess(@Observes(during=TransactionPhase.AFTER_SUCCESS) CourseStaffMemberUpdatedEvent event) {
+    List<Long> updatedCourseStaffMemberIds = datas.retrieveUpdatedCourseStaffMemberIds();
+    
+    for (Long updatedCourseStaffMemberId : updatedCourseStaffMemberIds) {
+      Long courseId = datas.getCourseStaffMemberCourseId(updatedCourseStaffMemberId);
+      Long staffMemberId = datas.getCourseStaffMemberStaffMemberId(updatedCourseStaffMemberId);
+      webhooks.sendWebhookNotification(new WebhookCourseStaffMemberUpdatePayload(updatedCourseStaffMemberId, courseId, staffMemberId));
+    }
+    
+    datas.clearUpdatedCourseStaffMemberIds();
```
Comment (ilmoeuro) [inline pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java]: What should happen if a webhook notification fails? Should this be cleared partially, totally, or not at all?

File: pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+  public synchronized void onCourseStudentUpdatedAfterSuccess(@Observes(during=TransactionPhase.AFTER_SUCCESS) CourseStudentUpdatedEvent event) {
+    List<Long> updatedCourseStudentIds = datas.retrieveUpdatedCourseStudentIds();
+    
+    for (Long updatedCourseStudentId : updatedCourseStudentIds) {
+      Long courseId = datas.getCourseStudentCourseId(updatedCourseStudentId);
+      Long studentId = datas.getCourseStudentStudentId(updatedCourseStudentId);
+      webhooks.sendWebhookNotification(new WebhookCourseStudentUpdatePayload(updatedCourseStudentId, courseId, studentId));
+    }
+    
+    datas.clearUpdatedCourseStudentIds();
```
Comment (ilmoeuro) [inline pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java]: See above comment.

File: pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/Webhooks.java [hunk 3]
```diff
@@ -1,272 +1,102 @@
...
-  
-  public synchronized void onStudentUpdatedBeforeCompletion(@Observes(during=TransactionPhase.BEFORE_COMPLETION) StudentUpdatedEvent event) {
-    datas.addUpdatedStudentId(event.getStudentId());
-  }
+  private ExecutorService executorService;
+  private List<fi.otavanopisto.pyramus.webhooks.Webhook> webhooks;
 
-  public synchronized void onStudentUpdatedAfterSuccess(@Observes(during=TransactionPhase.AFTER_SUCCESS) StudentUpdatedEvent event) {
-    List<Long> updatedStudentIds = datas.retrieveUpdatedStudentIds();
+  private class NotificationCallable implements Callable<Boolean> {
```
Comment (ilmoeuro) [inline pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/Webhooks.java]: Could this just be a method, instead of an inner class?

Comment (ilmoeuro) [reply pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/Webhooks.java]: I mean like this:

    FutureTask<Boolean> futureTask = new FutureTask(() -> notify(webhook.getUrl(), webhook.getSignature(), data);
    /* ... */
    Boolean notify(String url, String signature, String data) {
        /* ... */
    }

File: pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java (hunk 1 above)
Comment (anttiviljakainen) [reply pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java]: This is old code as-is. There is a separate issue where these should be covered: https://github.com/otavanopisto/pyramus/issues/536.

File: pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java (hunk 2 above)
Comment (anttiviljakainen) [reply pyramus/src/main/java/fi/otavanopisto/pyramus/webhooks/WebhookEventListeners.java]: This is old code as-is. There is a separate issue where these should be covered: https://github.com/otavanopisto/pyramus/issues/536.

**Pull request merged**